executor/common_zlib.h

```diff
@@ -0,0 +1,540 @@
+	for (len = 1; len <= MAXBITS; len++) {
+		left <<= 1; // one more bit, double codes left
+		left -= h->count[len]; // deduct count from possible codes
+		if (left < 0)
+			return left; // over-subscribed--return negative
+	} // left > 0 means incomplete
+
+	// generate offsets into symbol table for each length for sorting
+	offs[1] = 0;
+	for (len = 1; len < MAXBITS; len++)
+		offs[len + 1] = offs[len] + h->count[len];
+
+	// put symbols in table sorted by length, by symbol order within each length
+	for (symbol = 0; symbol < n; symbol++)
+		if (length[symbol] != 0)
+			h->symbol[offs[length[symbol]]++] = symbol;
+
+	// return zero for complete set, positive for incomplete set
+	return left;
+}
+
+// Decode literal/length and distance codes until an end-of-block code.
+static int puff_codes(struct puff_state* s,
+		      const struct puff_huffman* lencode,
+		      const struct puff_huffman* distcode)
+{
+	int symbol; // decoded symbol
+	int len; // length for copy
+	unsigned dist; // distance for copy
+	static const short lens[29] = {// Size base for length codes 257..285
+				       3, 4, 5, 6, 7, 8, 9, 10, 11, 13, 15, 17, 19, 23, 27, 31,
+				       35, 43, 51, 59, 67, 83, 99, 115, 131, 163, 195, 227, 258};
+	static const short lext[29] = {// Extra bits for length codes 257..285
+				       0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2,
+				       3, 3, 3, 3, 4, 4, 4, 4, 5, 5, 5, 5, 0};
+	static const short dists[30] = {// Offset base for distance codes 0..29
+					1, 2, 3, 4, 5, 7, 9, 13, 17, 25, 33, 49, 65, 97, 129, 193,
+					257, 385, 513, 769, 1025, 1537, 2049, 3073, 4097, 6145,
+					8193, 12289, 16385, 24577};
+	static const short dext[30] = {// Extra bits for distance codes 0..29
+				       0, 0, 0, 0, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6,
+				       7, 7, 8, 8, 9, 9, 10, 10, 11, 11,
+				       12, 12, 13, 13};
+
+	// decode literals and length/distance pairs
+	do {
+		symbol = puff_decode(s, lencode);
+		if (symbol < 0)
+			return symbol; // invalid symbol
+		if (symbol < 256) { // literal: symbol is the byte
+			// write out the literal
+			if (s->out != (unsigned char*)0) {
+				if (s->outcnt == s->outlen)
+					return 1;
+				s->out[s->outcnt] = symbol;
+			}
+			s->outcnt++;
+		} else if (symbol > 256) { // length
+			// get and compute length
+			symbol -= 257;
+			if (symbol >= 29)
+				return -10; // invalid fixed code
+			len = lens[symbol] + puff_bits(s, lext[symbol]);
+
+			// get and check distance
+			symbol = puff_decode(s, distcode);
+			if (symbol < 0)
+				return symbol; // invalid symbol
+			dist = dists[symbol] + puff_bits(s, dext[symbol]);
+#ifndef INFLATE_ALLOW_INVALID_DISTANCE_TOOFAR_ARRR
+			if (dist > s->outcnt)
+				return -11; // distance too far back
+#endif
+
+			// copy length bytes from distance bytes back
+			if (s->out != (unsigned char*)0) {
+				if (s->outcnt + len > s->outlen)
+					return 1;
+				while (len--) {
+					s->out[s->outcnt] =
+#ifdef INFLATE_ALLOW_INVALID_DISTANCE_TOOFAR_ARRR
```

Comment:
Let's delete the ifdef (but keep the code), it seems to harden the logic in a useful way.